### PR TITLE
[DROOLS-6341] Create codegen-implementation of Scorecard model

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @kostola @mariofusco @jiripetrlik

See https://issues.redhat.com/browse/DROOLS-6341

drools/kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard module generated

Depends on kiegroup/droolsjbpm-build-bootstrap#1644

Depends on DROOLS-6217

### CODE REVIEW

**Comment by gitgabrio** [comment]
Jenkins run cdb

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/src/test/java/org/kie/pmml/models/scorecard/tests/SimpleScorecardWithTransformationsTest.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        Assertions.assertThat(pmml4Result.getResultVariables().get(TARGET_FIELD)).isNotNull();
+        Assertions.assertThat(pmml4Result.getResultVariables().get(TARGET_FIELD)).isEqualTo(score);
+
+        Assertions.assertThat(pmml4Result.getResultVariables().get(REASON_CODE1_FIELD)).isEqualTo(reasonCode1);
+        Assertions.assertThat(pmml4Result.getResultVariables().get(REASON_CODE2_FIELD)).isEqualTo(reasonCode2);
+
+        // TODO {gcardosi} TO BE FIXED WITH DROOLS-5490
+//        Assertions.assertThat(pmml4Result.getResultVariables().get(OUT_DER_INPUT1)).isEqualTo(input1);
+//        Assertions.assertThat(pmml4Result.getResultVariables().get(OUT_DER_INPUT2)).isEqualTo(input2);
+//        Assertions.assertThat(pmml4Result.getResultVariables().get(OUT_DER_CONSTANT)).isEqualTo(CONSTANT);
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/src/test/java/org/kie/pmml/models/scorecard/tests/SimpleScorecardWithTransformationsTest.java]: This should be solved (DROOLS-5490 is resolved) so can you remove the comments and enable the code?

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/src/test/java/org/kie/pmml/models/scorecard/tests/MultipleCompoundNestedPredicateScorecardTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        inputData.put("input1", input1);
+        inputData.put("input2", input2);
+        PMML4Result pmml4Result = evaluate(pmmlRuntime, inputData, MODEL_NAME);
+        Assertions.assertThat(pmml4Result).isNotNull();
+
+        //  Currently unsupported for drools models
+//        Assertions.assertThat(pmml4Result.getResultVariables().get(TARGET_FIELD)).isNotNull();
+//        Assertions.assertThat(pmml4Result.getResultVariables().get(TARGET_FIELD)).isEqualTo(score);
+//        Assertions.assertThat(pmml4Result.getResultVariables().get(REASON_CODE1_FIELD)).isEqualTo(reasonCode1);
+//        Assertions.assertThat(pmml4Result.getResultVariables().get(REASON_CODE2_FIELD)).isEqualTo(reasonCode2);
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/src/test/java/org/kie/pmml/models/scorecard/tests/MultipleCompoundNestedPredicateScorecardTest.java]: This is not anymore a drools model, can you enable it?

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-model/src/main/java/org/kie/pmml/models/scorecard/model/KiePMMLScorecardModel.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+    @Override
+    public Object evaluate(final Object knowledgeBase, final Map<String, Object> requestData) {
+        return characteristicsFunction.apply(requestData, outputFieldsMap);
+    }
+
+
+
+
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-model/src/main/java/org/kie/pmml/models/scorecard/model/KiePMMLScorecardModel.java]: Too many empty lines

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-compiler/src/main/java/org/kie/pmml/models/scorecard/compiler/factories/KiePMMLCharacteristicsFactory.java [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+        final CompilationUnit characteristicsCloneCU =
+                JavaParserUtils.getKiePMMLModelCompilationUnit(containerClassName,
+                                                                                                      packageName,
+                                                                                                      KIE_PMML_CHARACTERISTICS_TEMPLATE_JAVA, KIE_PMML_CHARACTERISTICS_TEMPLATE);
+        final ClassOrInterfaceDeclaration characteristicsTemplate =
+                characteristicsCloneCU.getClassByName(containerClassName)
+                .orElseThrow(() -> new KiePMMLException(MAIN_CLASS_NOT_FOUND + ": " + containerClassName));
+        final CompilationUnit characteristicCloneCU = JavaParserUtils.getKiePMMLModelCompilationUnit(containerClassName,
+                                                                                                     packageName,
+                                                                                                     KIE_PMML_CHARACTERISTIC_TEMPLATE_JAVA, KIE_PMML_CHARACTERISTIC_TEMPLATE);
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-compiler/src/main/java/org/kie/pmml/models/scorecard/compiler/factories/KiePMMLCharacteristicsFactory.java]: Fix code formatting

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-evaluator/src/main/java/org/kie/pmml/models/scorecard/evaluator/PMMLScorecardModelEvaluator.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                                resultCode = orderedReasonCodes.get(index);
+                            }
+                            toPopulate.updateResultVariable(resultVariableName, resultCode);
+                        }
+                        break;
+                    case PREDICTED_VALUE:
+                        toPopulate.updateResultVariable(outputField.getName(), toPopulate.getResultVariables().get(toPopulate.getResultObjectName()));
+                        break;
+                    default:
+                        // not managed, yet
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-evaluator/src/main/java/org/kie/pmml/models/scorecard/evaluator/PMMLScorecardModelEvaluator.java]: Ticket id? Exception?

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-compiler/src/main/java/org/kie/pmml/models/scorecard/compiler/factories/KiePMMLCharacteristicsFactory.java (hunk 4 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-compiler/src/main/java/org/kie/pmml/models/scorecard/compiler/factories/KiePMMLCharacteristicsFactory.java]: @danielezonca 
done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-model/src/main/java/org/kie/pmml/models/scorecard/model/KiePMMLScorecardModel.java (hunk 3 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-model/src/main/java/org/kie/pmml/models/scorecard/model/KiePMMLScorecardModel.java]: @danielezonca 
done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/src/test/java/org/kie/pmml/models/scorecard/tests/MultipleCompoundNestedPredicateScorecardTest.java (hunk 2 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/src/test/java/org/kie/pmml/models/scorecard/tests/MultipleCompoundNestedPredicateScorecardTest.java]: @danielezonca 
done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/src/test/java/org/kie/pmml/models/scorecard/tests/SimpleScorecardWithTransformationsTest.java (hunk 1 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/src/test/java/org/kie/pmml/models/scorecard/tests/SimpleScorecardWithTransformationsTest.java]: @danielezonca 
done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-evaluator/src/main/java/org/kie/pmml/models/scorecard/evaluator/PMMLScorecardModelEvaluator.java (hunk 5 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-evaluator/src/main/java/org/kie/pmml/models/scorecard/evaluator/PMMLScorecardModelEvaluator.java]: @danielezonca 
done

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/main/java/org/kie/pmml/evaluator/core/utils/PostProcess.java [hunk 6]
```diff
@@ -129,12 +131,27 @@ static void populateOutputFields(final KiePMMLModel model, final PMML4Result toU
                                                                            model,
                                                                            kiePMMLNameValues));
         }
+        List<KiePMMLOutputField> reasonCodeOutputFields = outputFieldsByFeature.get(RESULT_FEATURE.REASON_CODE);
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/main/java/org/kie/pmml/evaluator/core/utils/PostProcess.java:134]: Is reason code generic or specific to scorecard? Does it make sense to move to logic to common class?

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/main/java/org/kie/pmml/evaluator/core/utils/PostProcess.java [hunk 7]
```diff
@@ -129,12 +131,27 @@ static void populateOutputFields(final KiePMMLModel model, final PMML4Result toU
...
+        if (reasonCodeOutputFields != null) {
+            final Map<String, Double> sortedByValue
+                    = model.getOutputFieldsMap().entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue() instanceof Double && (Double) entry.getValue() > 0) // TODO: //
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/main/java/org/kie/pmml/evaluator/core/utils/PostProcess.java]: Todo? Leftover or new ticket?

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/main/java/org/kie/pmml/evaluator/core/utils/PostProcess.java (hunk 6 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/main/java/org/kie/pmml/evaluator/core/utils/PostProcess.java:134]: I want to keep it in common code to be available regardless on actual implementation of given model, not only on type

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/main/java/org/kie/pmml/evaluator/core/utils/PostProcess.java (hunk 7 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/main/java/org/kie/pmml/evaluator/core/utils/PostProcess.java]: @danielezonca 
Leftover - removed.

**Review by danielezonca — APPROVED** [review]

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/ExpressionFunctionUtils.java [hunk 8]
```diff
@@ -199,8 +199,8 @@ public static MethodDeclaration getExpressionMethodDeclarationWithKiePMMLNameVal
      * @return
      */
     public static MethodDeclaration getExpressionMethodDeclarationWithStringObjectMap(final Expression expression,
-                                                                                      final DataType dataType,
```
Comment (jiripetrlik) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/ExpressionFunctionUtils.java:202]: I would prefer to separate formatting changes to a separate PRs and not mix it with changes in functionality. It quite slow down a review. But let's leave it this way now.

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/pom.xml [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+
+  <dependencyManagement>
+    <!-- TODO MOVE TO kie-pmml-trusty/kie-pmml-dependencies/pom.xml -->
```
Comment (jiripetrlik) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/pom.xml]: I'm not sure but wasn't these methods moved somewhere upwards already? Correct me if I'm wrong but I see it here:
- https://github.com/kiegroup/droolsjbpm-build-bootstrap/blob/master/kie-pmml-bom/pom.xml#L155

**Review by jiripetrlik — APPROVED** [review]
I have two minor comments.

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/ExpressionFunctionUtils.java (hunk 8 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/ExpressionFunctionUtils.java:202]: :+1:

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/pom.xml (hunk 9 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-scorecard/kie-pmml-models-scorecard-tests/pom.xml]: Hi @jiripetrlik 
It is not exactly the same thing. Inside the bom (you pointed to) we only define the version of each artifact. Inside  `kie-pmml-trusty/kie-pmml-dependencies/pom.xml` we define the artifacts that would be inherited/include in projects that declares "kie-pmml-dependencies" as dependency. Said differently,  "kie-pmml-dependencies" define the default implementations we are providing:  I hope to have been clear. Anyway, I removed the comment, that was leftover misleading

**Comment by gitgabrio** [comment]
Jenkins retest this

**Comment by gitgabrio** [comment]
Jenkins run cdb
